drivers: pwm: extend API to support events

javlands opened the pull request:
- Extend the pwm API to also support pwm interrupts. 
- Implement pwm interrupts for the SAM controllers

javlands committed:
drivers: pwm: extend API to support events

Extend the PWM API to support events, as some
controllers allow interrupts if e.g., a pwm period
has ended or a fault occured.

Signed-off-by: Jaro Van Landschoot <jaro.vanlandschoot@basalte.be>

javlands committed:
drivers: pwm: sam: Add period and fault events

The SAM4S pwm supports several events. This commit
implements the events when a channel period has ended
or a fault event has occured.

Signed-off-by: Jaro Van Landschoot <jaro.vanlandschoot@basalte.be>

pdgendt reviewed (commented):
Update `tests/drivers/build_all/pwm/testcase.yaml` with:

```diff
diff --git a/tests/drivers/build_all/pwm/testcase.yaml b/tests/drivers/build_all/pwm/testcase.yaml
index 34156300b5d..270637d4625 100644
--- a/tests/drivers/build_all/pwm/testcase.yaml
+++ b/tests/drivers/build_all/pwm/testcase.yaml
@@ -43,6 +43,8 @@ tests:
     platform_allow:
       - sam_e70_xplained/same70q21
       - sam_v71_xult/samv71q21b
+    extra_configs:
+      - CONFIG_PWM_INTERRUPT=y
   drivers.pwm.stm32.build:
     platform_allow: disco_l475_iot1
   drivers.pwm.xec.build:
```

javlands committed:
tests: drivers: build_all: pwm: add pwm event to sam

Add the pwm event functions of the sam pwm driver to the
build all test.

Signed-off-by: Jaro Van Landschoot <jaro.vanlandschoot@basalte.be>

pdgendt commented:
Ping to all reviewers

Should this be brought to the architecture working group meeting?

pdgendt commented:
> I am not convinced this is the right approach. We typically do not just expose IRQs to the application like this.

What's the alternative? BTW, this is also very similar to the PWM capture API.

A usecase would be to use it for IR input/output, where I'd expect similar APIs. Envision a "remote control" subsystem using these drivers as carriers.

henrikbrixandersen commented:
> What's the alternative? BTW, this is also very similar to the PWM capture API.

I don't think it is. The PWM capture callback provides the captured values upon a complete capture. The API proposed here merely forwards an IRQ from the device to the upper layer (subsystem or application). 

> A usecase would be to use it for IR input/output, where I'd expect similar APIs. Envision a "remote control" subsystem using these drivers as carriers.

For that, wouldn't it be more abstract to offload the "train" of PWM cycles to be handled in the driver/device itself? E.g. "perform N PWM cycles with period X, pulse width Y with callback on completion", or even the same but with multiple encoded PWM configurations to be "played" like a stream"

With the API proposed here, the subsystem would receive a callback upon each completed PWM cycle and have to respond to that, preventing hardware offload for capable devices.

carlescufi commented:
Architecture WG meeting:

- The use case [here](https://github.com/zephyrproject-rtos/zephyr/pull/93275#issuecomment-3112366927) is discussed in the call
- The comment by @bjarki-andreasen [here](https://github.com/zephyrproject-rtos/zephyr/pull/93275#issuecomment-3131816910) raises no objections, seems like a good idea to everyone
- @carlescufi suggests a solution to the hardware acceleration issue raised by @henrikbrixandersen in [this comment](https://github.com/zephyrproject-rtos/zephyr/pull/93275#issuecomment-3108783671):
  - Start with the API in this PR (with the changes suggested by @bjarki-andreasen)
  - Later introduce a new function that allows the user to trigger a fixed set of N cycles (all with the same period and pulse):
     -  `int pwm_set_n_cycles(const struct device *dev, uint32_t channel, uint32_t period, uint32_t pulse, int n, pwm_flags_t flags);`
   - At the same time introduce a new event in the event mask that triggers a callback at the end of the N cycles programmed

ts4ling commented:
I get @henrikbrixandersen 's concerns ("preventing hardware offload") in a general view, but for my stepper use-case it doesn't matter. For my drv8849 stepper driver I'm using `k_usleep()` between every ramp up/down PWM-step and when I'm not ramping I just sleep longer. That's pretty ugly. I'd love to use a PWM interrupt to do some counting.

javlands committed:
samples: drivers: pwm: event: Add sample

Add a simple sample demonstrating how the PWM event API can be
used to precisely time 5 periods of a PWM signal.

Signed-off-by: Jaro Van Landschoot <jaro.vanlandschoot@basalte.be>

javlands commented:
Hi all

I've added a simple sample that changes the duty cycle from 50% to 0% and back every 5 periods

javlands committed:
doc: releases: 4.4: Add PWM events

Add an entry for the addition of PWM events

Signed-off-by: Jaro Van Landschoot <jaro.vanlandschoot@basalte.be>

nandojve commented:
> > Not from my side.
> 
> You still have a NAK on this PR.

Ok, than it is not a problem anymore : )

henrikbrixandersen commented on include/zephyr/drivers/pwm.h:
```diff
@@ -400,6 +419,52 @@ typedef void (*pwm_capture_callback_handler_t)(const struct device *dev,
+	uint32_t channel;
+
+	/** A mask of events the callback is interested in. */
+	pwm_flags_t event_mask;
```
I don't think we should be reusing the `pwm_flags_t` type for storing PWM events types. The flags are for channel configuration, not for event masks etc.

henrikbrixandersen reviewed (dismissed):
Thanks!
